### ISSUE TITLE
File Utilities.hh is missing and std=c++1y

### ISSUE THREAD

**Issue by ferdymercury** [VinzenzBildstein/LecroyScope2Root#1]
Thanks for your contribution!

File Utilities.hh is missing. Maybe it would be nice to add a README file, where it is specified that you need to clone also the CommandLineInterface repository. 

And when used with gcc 5, I needed to change -std=c++11 to -std=c++1y in the Makefile, otherwise error. Tested on Ubuntu 16 Mate. On both repositories I had to change the Makefile.

```
g++ Oscilloscope.cc -c -I/opt/root/include -ICommandLineInterface -I. -fPIC -pedantic -Wall -Wno-long-long -g -O3 -std=c++11 -o Oscilloscope.o
In file included from /usr/include/c++/5/experimental/string_view:39:0,
                 from /opt/root/include/RStringView.h:24,
                 from /opt/root/include/TString.h:41,
                 from /opt/root/include/TNamed.h:29,
                 from /opt/root/include/TGraph.h:25,
                 from /opt/root/include/TGraphErrors.h:25,
                 from Oscilloscope.hh:14,
                 from Oscilloscope.cc:1:
/usr/include/c++/5/bits/c++14_warning.h:32:2: error: #error This file requires compiler and library support for the forthcoming ISO C++ 2014 standard. This support is currently experimental, and must be enabled with the -std=c++1y or -std=gnu++1y compiler options.
 #error This file requires compiler and library support for the forthcoming \
  ^
Oscilloscope.cc:6:24: fatal error: Utilities.hh: No such file or directory
compilation terminated.
Makefile:55: recipe for target 'Oscilloscope.o' failed
make: *** [Oscilloscope.o] Error 1
```

By the way, you could post your contribution here:
https://root.cern.ch/phpBB3/viewforum.php?f=12

**Comment by VinzenzBildstein** [comment]
Thanks for the testing it. I will look into those issues. It compiles on CentOs without any issues using -std=c++11 and gcc4.9. This might be an issue in gcc5, so I'll try and test that. I'll also try and get the Utilities.hh include (probably via a sub-module).
However I'm quite busy right now, so this might take a bit.

**Comment by VinzenzBildstein** [comment]
Okay,  it has been too long since I last looked at this obviously. The submodule CommandLineInterface has already been included (which contains Utilities.hh). So I'm guessing this is an error in the Makefile because it seems that any dependency on libUtilities.so is missing.

**Comment by VinzenzBildstein** [comment]
I fixed one issue with the URL of the submodule. Currently you have to manually run two command after cloning:
git submodule init
git submodule update
It should be possible to simply put this into the Makefile.

**Comment by ferdymercury** [comment]
Thanks for the fast reply!

I have a techincal side comment about the library design, just suggestions or improvements, take it just a personal opinion (matter of taste) or comments for the future:
- If the goal is to have a one to one conversion from .trc to .root, then the voltages should be stored as int16_t instead of double. The user can later convert it.
- Also, storing one graph per function implies a significant overhead.
- I would like better one tree entry per waveform, not per filename. The division in filenames has normally no logical sense, just due to max-size issues.
- I would store the header as "basic types" instead as a custom class. It allows sharing the file with someone that has not installed a library. See example on how to store labels and numbers in a (sub)TDirectory called e.g. "Header". https://root.cern.ch/phpBB3/download/file.php?id=12836&mode=view https://root.cern.ch/phpBB3/viewtopic.php?t=20937#p100235

In summary:
File 1.trc 2.trc 3.trc --> output.root
Each sequence has e.g. 200 points times 1000 pulses
TFile
Header (TDirectory)
--- TNamed key-values
--- TVector values
Data (TTree)
branch1: int16_t VoltageArray[200]
for each pulse --> Fill

At the end you have 3000 pulses (entries) in your tree.

This structure allows easy drawing with e.g TTree::Draw("VoltageArray:$Iteration"). Regards.

**Comment by VinzenzBildstein** [comment]
Thanks for the suggestions. I wrote this program a long time ago, so I have to familiarize myself with it again. As to the suggestions you've made:
- As far as I know the data is stored as floating point values in the .trc file (at least the one I had were), so storing it as int16_t doesn't make much sense.
- I agree, storing the graph (in addition to the vectors that it is made of) creates some overhead, but it is the only way to quickly draw the waveforms as graphs (which can be very helpful to see how individual traces look like when plotting multiple graphs on top of each other). 
- Again, as far as I know, each file only contains a single waveform, so having one entry per filename is the same as having one entry per waveform.
- I'll look into storing the header as basic types instead of a custom class.

**Comment by ferdymercury** [comment]
Thanks.

- No, I think that data are stored as int8_t or int16_t. You get the floating point value just after multiplying with vertical gain and summing offset. See appendix II http://teledynelecroy.com/doc/docview.aspx?id=5891 and also your code: `fVoltage.push_back(fHeader.GetVerticalGain()*static_cast<double>(*(reinterpret_cast<int16_t*>(buffer.data()+i))) - fHeader.GetVerticalOffset()); `Also https://github.com/yetifrisstlama/readTrc https://github.com/williamhbell/LeCroyScopeDAQ https://de.mathworks.com/matlabcentral/fileexchange/26375-readlecroybinarywaveform-m  It should be sth like this, from the python script:
```
 #------------------------
        # Get binary format / endianess
        #------------------------
        if readX( fid, '?', wdOffset + 32 ):  #16 or 8 bit sample format?
            smplFmt = "int16"
        else:
            smplFmt = "int8"
        if readX( fid, '?', wdOffset + 34 ):  #Big or little endian?
            endi = "<"
        else:
endi = ">"
```

- In fact, to quickly draw the waveforms as graphs, if the data are stored as basic int16_t as C-arrays (e.g. the branch `int16_t* VoltageArray = new int16_t[200]`), one could just call `tree->Draw("VoltageArray:$Iteration","","",1,1`). Anyway,I understand that it may be practical to store this graph also in the TFile. Maybe an optional argument to control if this is stored would be useful. And, it should be a TH1 with number of bins == number of samples, rather than a TGraph, as the x points are consecutive and with fixed separation. This would also reduce overhead. Drawing on top of each other is equally easy as with TGraph.
- There are two measurement modi: the Single Trace and the Sequence Mode. In Sequence Mode, each file contains up to 1000 waveforms. They are just appended as bytes at the end, so you do not have to modify the program. This value (1000) is the SUBARRAY_COUNT of the header. So one could derive the number of data points per waveform by dividing `WAVE_ARRAY_COUNT/SUBARRAY_COUNT`, and reshape the data accordingly. Example: `WAVE_ARRAY_COUNT = 252000, SUBARRAY_COUNT=1000`. This means that 1000 pulses each a length of 252 points are stored one byte after the other. So one could define a` int16_t* VoltageArray = new int16_t[252]`, read the 252 first data points, fill the tree with this array, read next 252 points, fill, etc.


Thanks for the fast reply and regards.

**Comment by VinzenzBildstein** [comment]
- Okay, I checked the code, and you're right, for a binary file the data is stored as an integer. However this seems different in an ascii file where it is stored as a floating point number. And since the actual voltage is the desired end-result anyways, I apparently decided to just have one vector of double instead of two vector depending on whether it was a binary or ascii file that was read.
- The draw command you're giving creates a histogram, not a graph. The point of having graphs is that one can draw them with lines connecting the data points. This makes a difference if you plot multiple waveforms on top of each other. E.g. if you have waveforms that are either 1 for odd samples and 0 for even samples or vice versa, you couldn't see whether that is what's happening if you draw two of them or if you simply have one that is constant at 0 and one that is constant at 1. Like I said, in most cases this is not needed, but when one wants to check for such cases there is no way to draw multiple graphs on top of each other directly from the tree.
- You're right, apparently I'm ignoring the sequence mode completely (the WAVE_ARRAY_COUNT isn't even read out). It seems we only ever took data in single trace mode.

**Comment by ferdymercury** [comment]
-  Ok, I understand. So storing as float may be better than storing as double like now, concerning space storage?

- I disagree about the graphs vs histograms. I get your idea about the even-odd things, but you can also draw a histogram as a line instead of a bar. `h->Draw("L")` produces exactly the same as drawing a graph. So this way there would be no confusion.
